sigstore/_verify/policy.py

```diff
@@ -0,0 +1,158 @@
+
+# From: https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md
+_OIDC_ISSUER_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.1")
+_OIDC_GITHUB_WORKFLOW_TRIGGER_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.2")
+_OIDC_GITHUB_WORKFLOW_SHA_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.3")
+_OIDC_GITHUB_WORKFLOW_NAME_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.4")
+_OIDC_GITHUB_WORKFLOW_REPOSITORY_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.5")
+_OIDC_GITHUB_WORKFLOW_REF_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.6")
+_OTHERNAME_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.7")
+
+
+class VerificationResult(BaseModel):
+    success: bool
+
+    def __bool__(self) -> bool:
+        return self.success
+
+
+class VerificationSuccess(VerificationResult):
+    success: bool = True
+
+
+class VerificationFailure(VerificationResult):
+    success: bool = False
+    reason: str
+
+
+class VerificationPolicy(Protocol):
+    @abstractmethod
+    def verify(self, cert: Certificate) -> VerificationResult:
+        raise NotImplementedError
+
+
+class AnyOf:
+    """
+    The "any of" policy, corresponding to a logical OR between child policies.
+    """
+
+    def __init__(self, children: list[VerificationPolicy]):
+        self._children = children
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        verified = any(child.verify(cert) for child in self._children)
+        if verified:
+            return VerificationSuccess()
+        else:
+            return VerificationFailure(
+                reason=f"0 of {len(self._children)} policies succeeded"
+            )
+
+
+class _OIDCIssuer:
+    """
+    Verifies the certificate's OIDC issuer, identified by
+    an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.1`.
+
+    See: <https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726411--issuer>
+    """
+
+    def __init__(self, *, issuer: str):
+        self._issuer = issuer
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        # Check that the OIDC issuer extension is present, and contains the expected
+        # issuer string (which is probably a URL).
+        try:
+            oidc_issuer = cert.extensions.get_extension_for_oid(_OIDC_ISSUER_OID).value
+        except ExtensionNotFound:
+            return VerificationFailure(
+                reason="Certificate does not contain OIDC issuer extension"
+            )
+
+        # NOTE(ww): mypy is confused by the `Extension[ExtensionType]` returned
+        # by `get_extension_for_oid` above.
+        issuer_value = oidc_issuer.value.decode()  # type: ignore[attr-defined]
+        if issuer_value != self._issuer:
+            return VerificationFailure(
+                reason=(
+                    "Certificate's OIDC issuer does not match "
+                    f"(got {issuer_value}, expected {self._issuer})"
+                )
+            )
+
+        return VerificationSuccess()
+
+
+class Identity(_OIDCIssuer):
```

Comment:
Do we foresee a reason to check anything other than the SAN? I might be missing some context but I don't see why this needs to be some extensible and composable (the `AnyOf` thing especially).

Reply:
Yeah -- we might want to check the other certificate extensions provided by Sigstore, e.g. `1.3.6.1.4.1.57264.1.4` for the GitHub workflow name (per #157). For those we'll probably want to tie them to the issuer as well (specifically, they'll always be GitHub's token issuer), so I figured it made sense to make that part composable.

Reply:
I stubbed out the GitHub workflow claims, so those should now be a good example of checking things other than the SAN.